prime-router/src/main/kotlin/fhirengine/utils/FHIRBundleHelpers.kt

```diff
@@ -25,6 +31,86 @@ import java.util.stream.Stream
  * A collection of helper functions that modify an existing FHIR bundle.
  */
 const val conditionExtensionurl = "https://reportstream.cdc.gov/fhir/StructureDefinition/reportable-condition"
+const val conditionCodeExtensionURL = "https://reportstream.cdc.gov/fhir/StructureDefinition/condition-code"
+
+/**
+ * Looks up a condition code for the passed in [code] (typically a test) using the given [metadata] object
+ * @param code test (or other type) code to look up a condition for
+ * @param metadata metadata containing an observation-mapping lookup table
+ * @return the condition code or null if no code was found
+ */
+private fun lookupCondition(code: Coding, metadata: Metadata): Coding? {
+    val mappingTable = metadata.findLookupTable("observation-mapping").also {
+        if (it == null) { // could not load the table
+            throw IllegalStateException("Unable to load lookup table 'observation-mapping' for condition stamping")
+        }
+    }!!
+    val condition = mappingTable.caseSensitiveDataRowsMap.find { // search for the code
+        it[ObservationMappingConstants.TEST_CODE_KEY] == code.code
+    }
+    return if (condition.isNullOrEmpty()) { // could not find the code
+        null
+    } else { // code found; create Coding instance to return
+        Coding(
+            condition[ObservationMappingConstants.CONDITION_CODE_SYSTEM_KEY],
+            condition[ObservationMappingConstants.CONDITION_CODE_KEY],
+            condition[ObservationMappingConstants.CONDITION_NAME_KEY]
+        )
+    }
+}
+
+/**
+ * Retrieves loinc/snomed codes from [this] observation in known locations (code.coding and valueCodeableConcept.coding)
+ * @return a map of lists of codings keyed by their origin as a printable string
+ */
+fun Observation.getCodeSourcesMap(): Map<String, List<Coding>> {
+    val toReturn = mutableMapOf<String, List<Coding>>()
+    try {
+        toReturn[ObservationMappingConstants.BUNDLE_CODE_IDENTIFIER] = this.code.coding
+    } catch (error: FHIRException) {
+        if (error.message == null ||
+            !error.message!!.startsWith("Type mismatch: the type CodeableConcept was expected")
+            ) {
+            throw error
+        }
+    }
+    try {
+        toReturn[ObservationMappingConstants.BUNDLE_CODEABLE_IDENTIFIER] = this.valueCodeableConcept.coding
```

Comment:
Nit: looking at this again I think this might be a better name (lines up with the actual FHIR property)
```suggestion
        toReturn[ObservationMappingConstants.BUNDLE_VALUE_IDENTIFIER] = this.valueCodeableConcept.coding
```